### PR TITLE
(towards #1208) Move Directive nodes from psyGen to their own files

### PR DESCRIPTION
**Pull request by sergisiso**
To keep the reviewer and the git conflict resolver sane I haven't modified any code inside the moved blocks besides the necessary import statements and doctring references.

Unfortunately the boundary between the directives and PSyclone concepts is not as clean as I thought (e.g. directives reference InvokeSchedule, kernels, ..) but this is not tacked in this PR (in fact some references are in the gen_code and may go away soon anyways)

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter or @rupertford this is ready for review. The PR touches many lines but the changes are quite repetitive. The missing coverage lines are in the directives code that I moved, I will let the reviewer decide what I should do with them.

**Review by arporter — CHANGES_REQUESTED** [review]
Mammoth effort Sergi - many thanks for doing this. It feels really good to see psyGen.py shrinking.
I think ultimately we probably want a `psyir/nodes/directives` directory with separate classes in separate files (I note that the `acc_directives.py` is already 800 lines long although I guess it will shrink as we remove the gen_code stuff). However, I think this is a good step in that direction and I think it's important to get it on quickly as it has the potential for many conflicts :-) Therefore, I'm not requesting that you do that now.
All tests and examples are OK with compilation.
I can't find any outdated references to psyGen in the docs, tutorials, code or the tests (impressive).
Docs all build without new errors/warnings.
There are a few 'import outside top-level' warnings for the new acc_directive and omp_directive files, please could you fix.

File: src/psyclone/dynamo0p3.py [hunk 1]
```diff
@@ -5994,7 +5995,7 @@ class DynHaloExchangeStart(DynHaloExchange):
     :type vector_index: int
     :param parent: optional PSyIRe parent node (default None) of this \
     object
-    :type parent: :py:class:`psyclone.psyGen.node`
+    :type parent: :py:class:`psyclone.psyit.nodes.Node`
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/dynamo0p3.py:5998]: Fixed

File: src/psyclone/psyGen.py [hunk 2]
```diff
@@ -47,15 +47,14 @@
 import six
 from fparser.two import Fortran2003
 from psyclone.configuration import Config
-from psyclone.f2pygen import DirectiveGen, CommentGen, CallGen, PSyIRGen, \
-    UseGen
-from psyclone.core import AccessType, VariablesAccessInfo
+from psyclone.f2pygen import CommentGen, CallGen, PSyIRGen, UseGen
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py:50]: Done

File: src/psyclone/psyGen.py [hunk 3]
```diff
@@ -2767,6 +1490,7 @@ def reprod_reduction(self):
         do loop. If so report whether it has the reproducible flag
         set. Note, this also catches OMPParallelDo Directives but they
         have reprod set to False so it is OK.'''
+        from psyclone.psyir.nodes import OMPDoDirective
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py:1493]: I moved this up

File: src/psyclone/psyir/backend/c.py [hunk 4]
```diff
@@ -112,7 +112,7 @@ def assignment_node(self, node):
         PSyIR tree.
 
         :param node: An Assignment PSyIR node.
-        :type node: :py:class:`psyclone.psyGen.Assigment`
+        :type node: :py:class:`psyclone.psyir.nodes.Assigment`
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/c.py:115]: Done

File: src/psyclone/psyir/backend/sir.py [hunk 5]
```diff
@@ -290,7 +290,7 @@ def assignment_node(self, node):
         PSyIR tree.
 
         :param node: an Assignment PSyIR node.
-        :type node: :py:class:`psyclone.psyGen.Assigment`
+        :type node: :py:class:`psyclone.psyir.nodes.Assigment`
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/backend/sir.py:293]: Done

File: src/psyclone/psyir/nodes/acc_directives.py [hunk 6]
```diff
@@ -0,0 +1,820 @@
...
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
+#         I. Kavcic,    Met Office
+#         C.M. Maynard, Met Office / University of Reading
+#         J. Henrichs, Bureau of Meteorology
+# -----------------------------------------------------------------------------
+
+''' This module contains the OpenACC Directives node implementation.'''
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/acc_directives.py:40]: Done

File: src/psyclone/psyir/nodes/acc_directives.py [hunk 7]
```diff
@@ -0,0 +1,820 @@
...
+        return result
+
+    def end_string(self):
+        '''
+        :returns: the text for the end of this directive region.
+        :rtype: str
+
+        '''
+        # pylint: disable=no-self-use
+        return "acc end data"
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/acc_directives.py:820]: Done

File: src/psyclone/psyir/nodes/directive.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+        fp_parent.content.insert(ast_start_index, directive)
+
+        self.ast = directive
+        self.dir_body.ast = directive
+        # If this is a directive applied to a Loop then update the ast_end
+        # for this Node to point to the parse tree for the loop. We have to
+        # do this because the loop is a sibling (rather than a child) of the
+        # directive in the parse tree.
+        if not end_text and isinstance(first_child, Loop):
+            self.ast_end = fp_parent.content[ast_start_index+1]
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/directive.py:246]: Done

File: src/psyclone/psyir/nodes/omp_directives.py [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
+#         I. Kavcic,    Met Office
+#         C.M. Maynard, Met Office / University of Reading
+#         J. Henrichs, Bureau of Meteorology
+# -----------------------------------------------------------------------------
+
+''' This module contains the OpenMP Directives node implementation.'''
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/omp_directives.py:40]: Done

File: src/psyclone/psyir/nodes/omp_directives.py [hunk 10]
```diff
@@ -0,0 +1,617 @@
...
+
+from __future__ import absolute_import
+from psyclone.f2pygen import (AssignGen, UseGen, DeclGen, DirectiveGen,
+                              CommentGen)
+from psyclone.psyir.nodes.directive import Directive
+from psyclone.psyir.nodes.routine import Routine
+from psyclone.psyir.nodes.loop import Loop
+from psyclone.errors import GenerationError, InternalError
+from psyclone.configuration import Config
+from psyclone.core import AccessType, VariablesAccessInfo
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/omp_directives.py:50]: Done I think

File: src/psyclone/psyir/nodes/omp_directives.py [hunk 11]
```diff
@@ -0,0 +1,617 @@
...
+            raise GenerationError(
+                "An OpenMP PARALLEL DO can only be applied to a single loop "
+                "but this Node has {0} children: {1}".
+                format(len(self.dir_body.children), self.dir_body.children))
+
+        self._add_region(
+            start_text="parallel do default(shared), private({0}), "
+            "schedule({1})".format(",".join(self._get_private_list()),
+                                   self._omp_schedule),
+            end_text="end parallel do")
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/nodes/omp_directives.py:617]: Done :)

File: src/psyclone/psyir/transformations/psy_data_trans.py [hunk 12]
```diff
@@ -41,9 +41,9 @@
 from fparser.two.utils import walk
 from psyclone.configuration import Config
 from psyclone.nemo import NemoInvoke
-from psyclone.psyir.nodes import PSyDataNode, Schedule, Return
-from psyclone.psyGen import InvokeSchedule, OMPDoDirective, ACCDirective, \
-    ACCLoopDirective
+from psyclone.psyir.nodes import PSyDataNode, Schedule, Return, \
+    OMPDoDirective, ACCDirective, ACCLoopDirective
+from psyclone.psyGen import InvokeSchedule
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyir/transformations/psy_data_trans.py:46]: Done

File: src/psyclone/tests/nemo/transformations/openacc/kernels_directive_test.py [hunk 13]
```diff
@@ -43,7 +43,8 @@
 
 from fparser.common.readfortran import FortranStringReader
 
-from psyclone.psyGen import ACCKernelsDirective, Loop, PSyFactory
+from psyclone.psyGen import PSyFactory
+from psyclone.psyir.nodes import ACCKernelsDirective, Loop
 from psyclone.psyir.transformations import TransformationError, ProfileTrans
 from psyclone.transformations import ACCKernelsTrans, ACCLoopTrans
 from psyclone.psyir.nodes import Assignment
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/nemo/transformations/openacc/kernels_directive_test.py:50]: Done

File: src/psyclone/tests/nemo/transformations/openmp/openmp_test.py [hunk 14]
```diff
@@ -152,7 +151,7 @@ def test_omp_add_region_invalid_data_move():
 def test_omp_parallel_multi():
     ''' Check insertion of an OpenMP parallel region containing more than
     one node. '''
-    from psyclone.psyGen import OMPParallelDirective
+    from psyclone.psyir.nodes import OMPParallelDirective
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/nemo/transformations/openmp/openmp_test.py:154]: Done

File: src/psyclone/tests/nemo/transformations/openmp/openmp_test.py [hunk 15]
```diff
@@ -276,7 +275,7 @@ def test_omp_do_children_err():
     ''' Tests that we raise the expected error when an OpenMP parallel do
     directive has more than one child. '''
     from psyclone.transformations import OMPParallelLoopTrans
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/nemo/transformations/openmp/openmp_test.py:277]: Done

File: src/psyclone/tests/psyGen_test.py [hunk 16]
```diff
@@ -35,13 +35,11 @@
...
 
-''' Performs py.test tests on the psygen module '''
+''' Performs py.test tests on the psyGen module '''
 
 
 # internal classes requiring tests
-# PSy,Invokes,Dependencies,Invoke,Node,Schedule,
-# LoopDirective,OMPLoopDirective,Loop,Call,Inf,SetInfCall,Kern,Arguments,
-# InfArguments,Argument,KernelArgument,InfArgument
+# PSy,Invokes,Invoke, Kern,Arguments, Argument, KernelArgument
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyGen_test.py:42]: Done

**Comment by sergisiso** [comment]
@arporter All comments addressed, this is ready for the next review.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
All examples and tests are still OK with compilation.
Although some of the moved code is not covered, that's not the fault of this PR and as it's only moved I think we keep things simple and live with the lack of coverage.
Updated reference guide builds without new warnings.
This PR does not close #1208 so I've not searched the source for any refs to it.

**Pull request merged**